buildpacks/metrics-agent/src/main.rs

```diff
@@ -0,0 +1,114 @@
+mod layers;
+
+use crate::layers::install_agentmon::{DownloadAgentmonError, InstallAgentmon};
+use commons::build_output::{self, fmt::ErrorInfo};
+use indoc::formatdoc;
+use libcnb::{
+    build::{BuildContext, BuildResult, BuildResultBuilder},
+    buildpack_main,
+    data::{build_plan::BuildPlanBuilder, layer_name},
+    detect::{DetectContext, DetectResult, DetectResultBuilder},
+    Buildpack,
+};
+
+#[derive(Debug)]
+enum MetricsAgentError {
+    DownloadAgentmonError(DownloadAgentmonError),
+}
+
+impl From<MetricsAgentError> for libcnb::Error<MetricsAgentError> {
+    fn from(error: MetricsAgentError) -> Self {
+        libcnb::Error::BuildpackError(error)
+    }
+}
+
+buildpack_main!(MetricsAgentBuildpack);
+
+pub(crate) struct MetricsAgentBuildpack;
+
+impl Buildpack for MetricsAgentBuildpack {
+    type Platform = libcnb::generic::GenericPlatform;
+    type Metadata = libcnb::generic::GenericMetadata;
+    type Error = MetricsAgentError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        let plan_builder = BuildPlanBuilder::new().provides("heroku-statsd-metrics-agent");
+
+        if let Ok(true) = fs_err::read_to_string(context.app_dir.join("Gemfile.lock"))
+            .map(|lockfile| lockfile.contains("barnes"))
+        {
+            DetectResultBuilder::pass()
+                .build_plan(plan_builder.requires("heroku-statsd-metrics-agent").build())
+                .build()
+        } else {
+            DetectResultBuilder::pass()
+                .build_plan(plan_builder.build())
+                .build()
+        }
+    }
+
+    fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
+        let build_duration = build_output::buildpack_name("Heroku StatsD Metrics Agent");
+
+        let section = build_output::section("Metrics agent");
+        context.handle_layer(
+            layer_name!("statsd-metrics-agent"),
+            InstallAgentmon { section },
+        )?;
+
+        build_duration.done_timed();
+        BuildResultBuilder::new().build()
+    }
+
+    fn on_error(&self, err: libcnb::Error<Self::Error>) {
+        on_error(err);
+    }
+}
+
+#[derive(Debug)]
+enum Cause {
+    OurError(MetricsAgentError),
+    FrameworkError(libcnb::Error<MetricsAgentError>),
+}
+
+fn cause(err: libcnb::Error<MetricsAgentError>) -> Cause {
+    match err {
+        libcnb::Error::BuildpackError(err) => Cause::OurError(err),
+        err => Cause::FrameworkError(err),
+    }
+}
+
+pub(crate) fn on_error(err: libcnb::Error<MetricsAgentError>) {
+    match cause(err) {
+        Cause::OurError(error) => log_our_error(error),
+        Cause::FrameworkError(error) => ErrorInfo::header_body_details(
+            "heroku/buildpack-ruby internal buildpack error",
```

Comment:
I don't understand why the additional `Cause` wrapper needed to be introduced here, the following seems to work just fine:

```suggestion
pub(crate) fn on_error(err: libcnb::Error<MetricsAgentError>) {
    match err {
        libcnb::Error::BuildpackError(error) => log_our_error(error),
        error => ErrorInfo::header_body_details(
            "heroku/buildpack-ruby internal buildpack error",
```

Maybe even better: using `libherokubuildpack::error::on_error`. If that is not good enough (the error message here is much more elaborate) we should change it there and let all other buildpacks reap the benefits too. :)